Fix handling of duplicates for `replace` on has_many-through

febeling opened the pull request:
There is a bug in the handling of duplicates when assigning (replacing) associated records, which made the result dependent on whether a given record was associated already, before being assigned anew. E.g.

    post.people = [person, person]
    post.people.count
    # => 2

while

    post.people = [person]
    post.people = [person, person]
    post.people.count
    # => 1

This change adds a test to provoke the former incorrect behavior, and fixes it.

Cause of the bug was the handling of record collections as sets, and using `-` (difference) and `&` (intersection) operations on them indiscriminately. This temporary conversion to sets would eliminate duplicates.

The fix calculates an occurrence distribution hash, with counts for each element. Based on these counts items are kept or removed in the difference and intersection operations.

Fixes #33942.

febeling committed:
Fix handling of duplicates for `replace` on has_many-through

There was a bug in the handling of duplicates when
assigning (replacing) associated records, which made the result
dependent on whether a given record was associated already before
being assigned anew. E.g.

    post.people = [person, person]
    post.people.count
    # => 2

while

    post.people = [person]
    post.people = [person, person]
    post.people.count
    # => 1

This change adds a test to provoke the former incorrect behavior, and
fixes it.

Cause of the bug was the handling of record collections as sets, and
using `-` (difference) and `&` (union) operations on them
indiscriminately. This temporary conversion to sets would eliminate
duplicates.

The fix is to decorate record collections for these operations, and
only for the `has_many :through` case. It is done by counting
occurrences, and use the record together with the occurrence number as
element, in order to make them work well in sets. Given

    a, b = *Person.all

then the collection used for finding the difference or union of
records would be internally changed from

     [a, b, a]

to

     [[a, 1], [b, 1], [a, 2]]

for these operations. So a first occurrence and a second occurrence
would be distinguishable, which is all that is necessary for this
task.

Fixes #33942.

febeling committed:
Rename union to intersection

febeling commented:
Just fixed the naming mixup between union and intersection.

petestreet commented:
Hi @febeling! This is a great find. There were a couple things I noticed that might be improved: 

- typo in `occurences` method name :)
- not sure about the "from_set" / "as_set" terminology, as these methods are dealing with arrays and hashes rather than actual Sets. 

I also took inspiration from [this stackoverflow answer](https://stackoverflow.com/a/24989593), which accomplishes the same work that this PR is doing with the `map` and `zip` operations, but a little more succinctly. A potential implementation would be to modify the code added to the [`has_many_through_association.rb`](https://github.com/rails/rails/pull/33954/commits/76f9b5ece6ad833351275445ab9f9d1c886dc293#diff-faefed8fb304ae45721a2645df0e1789R166) file to the following: 

```ruby
def difference(a, b)
  counts_b = occurrences(b)

  a.reject do |object|
    occurrence?(counts_b, object)
  end
end

def intersection(a, b)
  counts_b = occurrences(b)

  a.select do |object|
    occurrence?(counts_b, object)
  end
end

def occurrences(array)
  array.each_with_object(Hash.new(0)) do |object, counts|
    counts[object] += 1
  end
end

def occurrence?(counts, object)
  counts[object] > 0 && counts[object] -= 1
end
```

I ran benchmarks of the original version against the modified code in [this gist](https://gist.github.com/petestreet/169edbceccfba44f31860fd6ad184a61). The speed of execution is comparable (within a ~5% margin of error), but this modified code uses roughly 5x less memory.

febeling committed:
Optimize difference and intersection

febeling commented:
Thanks for your thorough review, @petestreet. The optimization you have researched looks absolutely valid, and I've changed the code to apply it. I also changed the method names to some I find more descriptive of the now changed algorithm. I hope you follow with these. What do you think?

petestreet commented:
Looks good, @febeling. Potentially would want to have comments indicating that `mark_occurrence` and `distribution` go together with `difference`/`intersection`, but Rails core members might have better opinions on style there.

febeling commented:
I realize the methods `intersection` and `difference` are a bit alien entities in in this class. They might be well-located in an Array extension. Because they have little connection with the rest of the AR module, I also abstained from adding tests for them (which would be alien to the testsuite) - something that otherwise seems called for, given the complexity of the implementation.

So I'd love to hear an opinion by a core member if, firstly, this should move over to, say activesupport/core_ext/array. I'm sure there are comparable cases where code is general in applicability, but rarely used, and if that's considered good material for AS - yes or no?

Secondly, what can be done to help this get merged? I do realize this is a lot of code for a bug fix. The reason for that is, that duplicates weren't actually using the correct operators, meaning the implementation was actually _missing_, for a very narrow edge case, to be fair (partial replacement _while also_ using duplicates - something that's uncommon generally, and not even possible with regular has_many).

febeling commented:
Just added very basic test examples for illustration of `intersection` and `difference` for arrays in a separate commit.

febeling commented:
@georgeclaghorn Any thoughts on this fix?

foucist commented:
How about defining & and - methods directly in ActiveRecord::Associations::HasManyThroughAssociation rather than introducing the intersection/difference methods?

tenderlove commented on activerecord/test/cases/associations/has_many_through_associations_test.rb:
```diff
@@ -586,6 +586,31 @@ def test_replace_association
+    assert_equal association.send(:difference,
+       [1, 1, 2, 3, 3],
+       [1, 3, 3, 3, 4]
+    ), [1, 2]
```
I'd prefer we don't use `send` on an internal API.  Is the above integration test not enough to prevent regression?  If not, can we add another integration test?  That way we are free to change the internal API later.

febeling replied:
Thanks for your review!

Yes, good point. I looked at the code more, and I think we're good with only the integration test.

I tossed the respective commit and pushed again.

tenderlove reviewed (commented):
The implementation in this PR looks good to me, but can we change the test around a little?  Thanks!

febeling commented:
@tenderlove I think the concerns are addressed, wdyt?

tenderlove reviewed (approved):
Looks great! Sorry for the delayed feedback.

febeling commented:
@tenderlove thanks!

febeling commented:
@tenderlove Actually, I’m really stoked about suddenly becoming a Rails contributor 🎉😁
